File: scala-spring/transactions-command-side-backend/src/main/scala/net/chrisrichardson/eventstore/examples/bank/transactions/MoneyTransfer.scala
package net.chrisrichardson.eventstore.examples.bank.transactions

import net.chrisrichardson.eventstore.PatternMatchingCommandProcessingAggregate
import net.chrisrichardson.eventstore.examples.bank.backend.common.transactions._
import net.chrisrichardson.eventstore.examples.bank.transactions.MoneyTransferCommands._

object TransferStates {
  sealed trait State
  object NEW extends State
  object INITIAL extends State
  object DEBITED extends State
  object COMPLETED extends State
  object FAILED_DUE_TO_INSUFFICIENT_FUNDS extends State
}



case class MoneyTransfer(state : TransferStates.State, details : TransferDetails)
  extends PatternMatchingCommandProcessingAggregate[MoneyTransfer, MoneyTransferCommand] {

  def this() = this(TransferStates.NEW, null)

  def processCommand = {
    case CreateMoneyTransferCommand(withDetails) if state == TransferStates.NEW =>
      Seq(MoneyTransferCreatedEvent(withDetails))

    case RecordDebitCommand(accountId) if state == TransferStates.INITIAL && accountId == details.fromAccountId =>
      Seq(DebitRecordedEvent(details))

    case RecordDebitFailedDueToInsufficientFundsCommand(accountId) if state == TransferStates.INITIAL && accountId == details.fromAccountId =>
      Seq(TransferFailedDueToInsufficientFundsEvent())

    case RecordCreditCommand(accountId) if state == TransferStates.DEBITED && accountId == details.toAccountId =>
      Seq(CreditRecordedEvent(details))
  }

  def applyEvent = {

    case MoneyTransferCreatedEvent(withDetails) => copy(state=TransferStates.INITIAL, details=withDetails)
    case DebitRecordedEvent(_) => copy(state=TransferStates.DEBITED)
    case TransferFailedDueToInsufficientFundsEvent() => copy(state=TransferStates.FAILED_DUE_TO_INSUFFICIENT_FUNDS)
    case CreditRecordedEvent(_) => copy(state=TransferStates.COMPLETED)

  }

}

File: java-spring/transactions-service/src/main/java/net/chrisrichardson/eventstore/javaexamples/banking/transactionsservice/backend/MoneyTransfer.java
package net.chrisrichardson.eventstore.javaexamples.banking.transactionsservice.backend;

import io.eventuate.Event;
import io.eventuate.EventUtil;
import io.eventuate.ReflectiveMutableCommandProcessingAggregate;
import net.chrisrichardson.eventstore.javaexamples.banking.backend.common.transactions.*;
import net.chrisrichardson.eventstore.javaexamples.banking.common.transactions.TransferState;

import java.util.List;

public class MoneyTransfer extends ReflectiveMutableCommandProcessingAggregate<MoneyTransfer, MoneyTransferCommand> {

  private TransferDetails details;
  private TransferState state;

  public List<Event> process(CreateMoneyTransferCommand cmd) {
    return EventUtil.events(new MoneyTransferCreatedEvent(cmd.getDetails()));
  }

  public List<Event> process(RecordDebitCommand cmd) {
    return EventUtil.events(new DebitRecordedEvent(details));
  }

  public List<Event> process(RecordDebitFailedCommand cmd) {
    return EventUtil.events(new FailedDebitRecordedEvent(details));
  }

  public List<Event> process(RecordCreditCommand cmd) {
    return EventUtil.events(new CreditRecordedEvent(details));
  }

  public void apply(MoneyTransferCreatedEvent event) {
    this.details = event.getDetails();
    this.state = TransferState.INITIAL;
  }

  public void apply(DebitRecordedEvent event) {
    this.state = TransferState.DEBITED;
  }

  public void apply(FailedDebitRecordedEvent event) {
    this.state = TransferState.FAILED_DUE_TO_INSUFFICIENT_FUNDS;
  }

  public void apply(CreditRecordedEvent event) {
    this.state = TransferState.COMPLETED;
  }


  public TransferState getState() {
    return state;
  }
}
